pkg/security/admission/admission.go

```diff
@@ -115,20 +116,55 @@ func (c *constraint) Admit(a admission.Attributes) error {
 		return admission.NewForbidden(a, fmt.Errorf("no providers available to validate pod request"))
 	}
 
+	// TODO(liggitt): allow spec mutation during initializing updates?
+	specMutationAllowed := a.GetOperation() == admission.Create
+
 	// all containers in a single pod must validate under a single provider or we will reject the request
 	validationErrs := field.ErrorList{}
+	var (
+		allowedPod       *kapi.Pod
+		allowingProvider scc.SecurityContextConstraintsProvider
+	)
+
+loop:
 	for _, provider := range providers {
-		if errs := scc.AssignSecurityContext(provider, pod, field.NewPath(fmt.Sprintf("provider %s: ", provider.GetSCCName()))); len(errs) > 0 {
+		clone, err := kapi.Scheme.DeepCopy(pod)
+		if err != nil {
+			return admission.NewForbidden(a, err)
+		}
+		podCopy := clone.(*kapi.Pod)
+
+		if errs := scc.AssignSecurityContext(provider, podCopy, field.NewPath(fmt.Sprintf("provider %s: ", provider.GetSCCName()))); len(errs) > 0 {
 			validationErrs = append(validationErrs, errs...)
 			continue
 		}
 
-		// the entire pod validated, annotate and accept the pod
-		glog.V(4).Infof("pod %s (generate: %s) validated against provider %s", pod.Name, pod.GenerateName, provider.GetSCCName())
+		// the entire pod validated
+		switch {
+		case apiequality.Semantic.DeepEqual(pod, podCopy):
+			// if it validated without mutating anything, use this result
+			allowedPod = podCopy
+			allowingProvider = provider
+			break loop
+		case specMutationAllowed && allowedPod == nil:
+			// if mutation is allowed and this is the first SCC to allow the pod, remember it,
```

Comment:
The approach that @liggitt has used for Kubernetes differs from the our current. For example, I created a simple pod under `system:admin` and a pod was admitted by the `privileged` SCC (because it accepts a pod without mutation). But the current version of OpenShift would admit my pod by the `anyuid` SCC (because it less permissive).

@liggitt @pweil- What is the desired behavior for this case? Preserve backward compatibility, be compatible with Kubernetes, or something else?

Reply:
@openshift/sig-security PTAL

> What is the desired behavior for this case? Preserve backward compatibility, be compatible with Kubernetes, or something else?

@liggitt @pweil- Do you have opinion on that?

Reply:
I think we have to preserve backwards compatibility as much as possible here while working around the mutation bug.  But if it doesn't mutate the pod does it matter which SCC admitted it?  I would think only a change in mutation behavior would be a problem here. (wouldn't `anyuid` still mutate the selinux labels where `privileged` would not?)

Reply:
on create, keep our current priority order. on update (where mutation is not allowed), follow the non-mutating logic.

Reply:
our handling of this will look different than PSP, because of our "prefer most restrictive" behavior we have to stay compatible with.

I'd expect something along these lines:
```go
// the entire pod validated
switch {
  case specMutationAllowed:
    // if we allow mutation, just use the first SCC that allows the pod
    allowedPod = podCopy
    allowingProvider = provider
    break loop

  case apiequality.Semantic.DeepEqual(pod, podCopy):
    // if we don't allow mutation, only use the validated pod if it didn't require any spec changes
    allowedPod = podCopy
    allowingProvider = provider
    glog.V(6).Infof("pod %s (generate: %s) validated against provider %s without mutation", pod.Name, pod.GenerateName, provider.GetSCCName()) 
    break loop

  default:
    glog.V(6).Infof("pod %s (generate: %s) validated against provider %s, but required mutation, skipping", pod.Name, pod.GenerateName, provider.GetSCCName())
}
```